src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs

```diff
@@ -0,0 +1,130 @@
+    {{
+        // Defined in 'PowerShell.Common.props' as 'ProductVersion'
+        // Example:
+        //  - when built from a commit:              ProductVersion = '7.3.0-preview.8 Commits: 29 SHA: 52c6b...'
+        //  - when built from a preview release tag: ProductVersion = '7.3.0-preview.8 SHA: f1ec9...'
+        //  - when built from a stable release tag:  ProductVersion = '7.3.0 SHA: f1ec9...'
+        internal const string ProductVersion = ""{0}"";
+
+        // The git commit id that the build is based off.
+        // Defined in 'PowerShell.Common.props' as 'PowerShellVersion' or 'ReleaseTag',
+        // depending on whether the '-ReleaseTag' is specified when building.
+        // Example:
+        //  - when built from a commit:              GitCommitId = '7.3.0-preview.8-29-g52c6b...'
+        //  - when built from a preview release tag: GitCommitId = '7.3.0-preview.8'
+        //  - when built from a stable release tag:  GitCommitId = '7.3.0'
+        internal const string GitCommitId = ""{1}"";
+
+        // The PowerShell version components.
+        // The version string is defined in 'PowerShell.Common.props' as 'PSCoreBuildVersion',
+        // but we break it into components to save the overhead of parsing at runtime.
+        // Example:
+        //  - '7.3.0-preview.8' for preview release or private build
+        //  - '7.3.0' for stable release
+        private const int Version_Major = {2};
+        private const int Version_Minor = {3};
+        private const int Version_Patch = {4};
+        private const string Version_Label = ""{5}"";
+    }}
+}}";
+
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.ProductVersion", out var productVersion);
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.PSCoreBuildVersion", out var mainVersion);
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.PowerShellVersion", out var gitDescribe);
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.ReleaseTag", out var releaseTag);
+
+            string gitCommitId = string.IsNullOrEmpty(releaseTag) ? gitDescribe : releaseTag;
+            if (gitCommitId.StartsWith("v"))
+            {
+                gitCommitId = gitCommitId.Substring(1);
+            }
+
+            var result = FastParsePSVersion(mainVersion);
+
+            return string.Format(
+                CultureInfo.InvariantCulture,
+                SourceTemplate,
+                productVersion,
+                gitCommitId,
+                result.major,
+                result.minor,
+                result.patch,
+                result.preReleaseLabel);
+        }
+
+        private static (int major, int minor, int patch, string preReleaseLabel) FastParsePSVersion(string mainVersion)
+        {
+            // We only handle the pre-defined PSVersion format here, e.g. 7.x.x or 7.x.x-preview.x
```

Comment:
We have no need to do "fast" on some ns since it is _build_ time.

Also, it is big headache to partially process versions in so many places (Build/Release.psm1, csproj, SG, SMA) - any minor change in one place can break all unpredictable.

_I strongly suggest keeping all version parsing in PowerShell.Common.props file._

As result, the SG will be very simple and stable, and our build process will be easily maintainable.

Reply:
> We have no need to do "fast" on some ns since it is build time.

It's just using the same function name. _fast_ is to compare with letting `SemanticVersion` do the string parsing.

> I strongly suggest keeping all version parsing in PowerShell.Common.props file.

I disagree. The reason to handle version in .csproj is that we need to declare `ProductVersion` for all assemblies. But only `S.M.A` needs the parsed components of the version string, so keeping it in `PowerShell.Common.props` will increase build time for all assemblies, unnecessarily.

This change is essentially moving the version parsing code from `SMA` to source generator (`PSVersionInfo` itself doesn't do version parsing anymore). If we need to change how PowerShell version will be represented in future (which I think is unlikely), then it's also easier to change the parsing in C# code and also easier to verify the change.

Reply:
> I disagree. The reason to handle version in .csproj is that we need to declare `ProductVersion` for all assemblies. But only `S.M.A` needs the parsed components of the version string, so keeping it in `PowerShell.Common.props` will increase build time for all assemblies, unnecessarily.

Handling version in .csproj probably takes a second or two. This is incomparable to the compile time which takes several minutes because of analyzers.

>If we need to change how PowerShell version will be represented in future (which I think is unlikely), then it's also easier to change the parsing in C# code and also easier to verify the change.

Exactly what is unlikely and this has already been concentrated in csproj for a long time. Now we are smearing it again.
When I opened this PR package tests were crashing because Release.psm1 was doing something incomprehensible. Now these tests are working, i.e. something was changed there again.
The release process is entirely on the MSFT team and above all you will waste your time over and over again. Although the community also releases their packages and they just benefit from having all the logic in csproj so they don't have to puzzle over the convoluted logic that is smeared out.

Reply:
> Handling version in .csproj probably takes a second or two. This is incomparable to the compile time which takes several minutes because of analyzers.

1. It's not only about a few more seconds, it's more about not doing things that are not needed. Again, all other assemblies other than `S.M.A` don't need the version to be parsed.
2. Parsing the version in C# code is more maintainable and easier to change if really needed.

> The release process is entirely on the MSFT team and above all you will waste your time over and over again

I don't know about other maintainers' experience regarding this, but I really cannot remember what the last time was I saw a build failed due to parsing the version string in C# code.

Reply:
If we would think about third-party linux repository maintainer they have to build from sources for security reasons. And they have to learn and port Build.psm1 and other helpers again and again. They must also support previous versions (security patches). It is obvious how difficult this is. If we add that last year git tags for some reason have problems all the time, they have a big headache. It would be much easier if it just worked `dotnet publish`.

Reply:
There are too many `if's` here :)
As far as I understand, it's not a goal as of today to build PowerShell completely without scripts due to the complexity of the PowerShell release build.

But, even if things were changed in future, parsing version in source generator code would not be a blocker to that goal -- it would work well with just `dotnet publish`.